File: src/main/java/org/liquibase/ext/persistence/utils/CommandExecutor.java
package org.liquibase.ext.persistence.utils;

import liquibase.ui.UIService;
import org.apache.commons.lang3.SystemUtils;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.TimeUnit;

public class CommandExecutor {

    private final UIService ui;
    private final Long duration;
    private final TimeUnit unit;

    public CommandExecutor(UIService ui, Long duration, TimeUnit unit) {
        this.ui = ui;
        this.duration = duration;
        this.unit = unit;
    }

    public CommandExecutor(UIService ui) {
        this.ui = ui;
        this.duration = 10L;
        this.unit = TimeUnit.MINUTES;
    }

    public void exec(List args) throws InterruptedException, IOException {

        ProcessBuilder builder = new ProcessBuilder();

        builder.command(args);
        Process process = builder.start();

        this.ui.sendMessage("----------------------------------------------------------------------");

        try {
            if (SystemUtils.IS_OS_MAC || SystemUtils.IS_OS_LINUX) {
                process.waitFor(this.duration, this.unit);
            }

            BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
            String line;
            while ( (line = reader.readLine()) != null) {
                this.ui.sendMessage(line);
                this.ui.sendMessage(System.getProperty("line.separator"));
            }

            if (process.isAlive()) {
                throw new IOException("Timed out waiting for command: " + args);
            }

            if (process.exitValue() != 0) {
                BufferedReader errReader = new BufferedReader(new InputStreamReader(process.getErrorStream()));
                String errLine;
                while ( (errLine = errReader.readLine()) != null) {
                    this.ui.sendErrorMessage(errLine);
                    this.ui.sendErrorMessage(System.getProperty("line.separator"));
                }
                throw new IOException("Command failed: " + this.stringify(args) + ": " + "");
            }
            if (SystemUtils.IS_OS_WINDOWS) {
                process.waitFor(this.duration, this.unit);
            }
        } finally {
            process.destroy();
        }
    }

    private String stringify(List $this$stringify) {
        String retString = "";
        String item;
        for(Iterator var4 = $this$stringify.iterator(); var4.hasNext(); retString = retString + item + ' ') {
            item = (String)var4.next();
        }
        return retString;
    }
}

File: src/main/kotlin/io/titandata/titan/utils/CommandExecutor.kt
/*
 * Copyright (c) 2021 by Titan Project Contributors.. All rights reserved.
 */

package io.titandata.titan.utils

import io.titandata.titan.exceptions.CommandException
import java.io.IOException
import java.util.concurrent.TimeUnit
import liquibase.Scope
import liquibase.ui.UIService
import org.apache.commons.lang3.SystemUtils

/**
 * ORIGINAL FILE FROM TITAN-SERVER
 *
 * Handle invocation of external commands. This is a wrapper around the native interfaces that
 * handles all of the stdin/stdout errors, and will throw an exception if the command returns
 * a non-zero exit status. It is also a convenient to mock out any dependencies on the external
 * system.
 */
class CommandExecutor(
    private val timeout: Long = 10,
    private val ui: UIService? = Scope.getCurrentScope().ui
) {

    /**
     * Execute the given command. Throws an exception if the
     */
    fun exec(args: List<String>, debug: Boolean = false): String {
        val builder = ProcessBuilder()
        if (debug) ui?.sendMessage(args.stringify())
        builder.command(args)
        val process = builder.start()
        try {
            /**
             Windows has an issue with the process.waitFor being at the beginning of this loop.
             It has been moved to the end with a condition special to Windows to catch this OS
             difference. Linux and MacOS want the process.waitFor at the beginning of the loop.
             */
            if (SystemUtils.IS_OS_MAC || SystemUtils.IS_OS_LINUX) {
                process.waitFor(timeout, TimeUnit.MINUTES)
            }
            val output = process.inputStream.bufferedReader().readText()
            if (process.isAlive) {
                throw IOException("Timed out waiting for command: $args")
            }
            if (debug) ui?.sendMessage("Exit Value: ${process.exitValue()}")
            if (process.exitValue() != 0) {
                val errOutput = process.errorStream.bufferedReader().readText()
                if (debug) ui?.sendMessage(errOutput)
                throw CommandException("Command failed: ${args.stringify()}",
                        exitCode = process.exitValue(),
                        output = errOutput)
            }
            if (SystemUtils.IS_OS_WINDOWS) {
                process.waitFor(timeout, TimeUnit.MINUTES)
            }
            return output
        } finally {
            process.destroy()
        }
    }

    companion object {
        fun List<String>.stringify(): String {
            var retString = ""
            for (item in this) {
                retString += "$item "
            }
            return retString
        }
    }
}
